main/src/ca/uwaterloo/flix/language/phase/unification/FastBoolUnification.scala

```diff
@@ -0,0 +1,1694 @@
+      debugln("-".repeat(80))
+      val s = propagateVars(currentEqns, currentSubst)
+      updateState(s)
+      printEquations()
+      printSubstitution()
+      debugln()
+    }
+
+    private def phase3VarAssignment(): Unit = {
+      debugln("-".repeat(80))
+      debugln("--- Phase 3: Variable Assignment")
+      debugln("    (resolves all equations of the form: x = t where x is free in t)")
+      debugln("-".repeat(80))
+      val s = varAssignment(currentEqns, currentSubst)
+      updateState(s)
+      printEquations()
+      printSubstitution()
+      debugln()
+    }
+
+    private def phase4SVE(): Unit = {
+      debugln("-".repeat(80))
+      debugln("--- Phase 4: Boolean Unification")
+      debugln("    (resolves all remaining equations using SVE.)")
+      debugln("-".repeat(80))
+      val newSubst = boolUnifyAllPickSmallest(currentEqns)
+      currentEqns = Nil
+      currentSubst = newSubst @@ currentSubst
+      printSubstitution()
+      debugln()
+    }
+
+    /**
+      * Updates the internal state with the given list of pending equations and partial substitution `l`.
+      *
+      * Simplifies the pending equations. Throws [[ConflictException]] if an unsolvable equation is detected.
+      */
+    private def updateState(l: (List[Equation], BoolSubstitution)): Unit = {
+      val (nextEqns, nextSubst) = l
+      currentEqns = simplify(nextEqns)
+      currentSubst = nextSubst
+    }
+
+    /**
+      * Verifies that the current substitution is a valid solution to the original equations `l`.
+      *
+      * Note: Does not make sense to call before the equation system has been fully solved.
+      *
+      * Throws a [[ConflictException]] if an equation is not solved by the current substitution.
+      */
+    private def verifySolution(): Unit = {
+      verify(currentSubst, l)
+    }
+
+    /**
+      * Checks that current substitution has not grown too large according to the defined threshold.
+      */
+    private def verifySolutionSize(): Unit = {
+      val size = currentSubst.size
+      if (size > Threshold) {
+        throw TooComplexException(size)
+      }
+    }
+
+    private def printEquations(): Unit = {
+      debugln(s"Equations (${currentEqns.size}):")
+      debugln(format(currentEqns))
+    }
+
+    private def printSubstitution(): Unit = {
+      debugln(s"Substitution (${currentSubst.numberOfBindings}):")
+      debugln(currentSubst.toString)
+    }
+
+    private def debugln(): Unit = debugln("")
+
+    private def debugln(s: String): Unit = {
+      if (Debugging) {
+        Console.println(s)
+      }
+    }
+
+  }
+
+  /**
+    * Returns a list of non-trivial unification equations computed from the given list `l`.
+    *
+    * Throws a [[ConflictException]] if an unsolvable equation is encountered.
+    *
+    * A trivial equation is one of:
+    * -     true ~ true
+    * -    false ~ false
+    * -        c ~ c        (same constant)
+    * -        x ~ x        (same variable)
```

Comment:
is it worth the linear run to check for syntactically equivalent formulas?

Reply:
Oh yes. SVE is sooo expensive.